modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -1225,6 +1232,82 @@ private void distributedExchange() throws IgniteCheckedException {
         initDone();
     }
 
+    /**
+     *
+     */
+    private void onDistributedExchangeTimeout() {
+        if (!isDone()) {
+            synchronized (mux) {
+                if (crd.isLocal()) {
+                    Set<UUID> remaining0 = new HashSet<>(this.remaining);
+
+                    log.warning("Exchange timeout reached on coordinator node, kicking pending nodes: " +
+                        S.arrayToString(UUID.class, remaining0.toArray()));
+
+                    for (UUID remainingUuid : remaining0)
+                        cctx.discovery().failNode(remainingUuid, null);
+                }
+                else {
+                    try {
+                        log.warning("Exchange timeout reached on non-coordinator node, " +
+                            "sending check request to coordinator");
+
+                        cctx.kernalContext().io().sendToGridTopic(
+                            crd, GridTopic.TOPIC_EXCHANGE, new PartitionsExchangeFinishedCheckRequest(), SYSTEM_POOL);
+
+                        cctx.time().schedule(() -> {
+                            if (!isDone())
+                                cctx.kernalContext().failure().process(
+                                    new FailureContext(FailureType.CRITICAL_ERROR,
+                                        new TimeoutException("Last exchange check response wasn't received.")));
+                        }, cctx.gridConfig().getExchangeHardTimeout(), -1);
+                    }
+                    catch (IgniteCheckedException ex) {
+                        cctx.kernalContext().failure().process(new FailureContext(FailureType.CRITICAL_ERROR, ex));
+
+                        throw new IgniteException(ex);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * @param lastFinishedVer Last finished topology version.
+     * @param pendingVer Pending topology version.
+     * @param receivedSingleMsg Received single message flag.
+     */
+    public void onCrdLastFinishedVersionReceived(AffinityTopologyVersion lastFinishedVer,
+        AffinityTopologyVersion pendingVer, boolean receivedSingleMsg) {
+        log.warning("Received check response from coordinator: " +
```

Comment:
The warning means that something bad happens, but from a user point of view, I can't understand what this message means and what actions are expected of me. I think we should move this to INFO level.

Reply:
If a coordinator didn't send a full message and stopped processing check requests, then all other nodes will die. I doubt, that this is a good outcome.